File: viewer/src/main/java/org/cryptobiotic/rlauxe/viewer/CardTable.java
/*
 * Copyright (c) 2025 John L. Caron
 * See LICENSE for license information.
 */

package org.cryptobiotic.rlauxe.viewer;

import org.cryptobiotic.rlauxe.audit.*;
import org.cryptobiotic.rlauxe.persist.*;
import org.cryptobiotic.rlauxe.workflow.PersistedMvrManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ucar.ui.prefs.BeanTable;
import ucar.ui.widget.IndependentWindow;
import ucar.ui.widget.TextHistoryPane;
import ucar.util.prefs.PreferencesExt;

import javax.swing.*;
import java.awt.*;
import java.util.*;
import java.util.List;

import static java.util.Collections.emptyList;

public class CardTable extends JPanel implements ViewerPanelIF {
    static private final Logger logger = LoggerFactory.getLogger(CardTable.class);

    private final PreferencesExt prefs;

    private final BeanTable<CardBean> cardTable;
    TextHistoryPane localInfo = new TextHistoryPane();

    private final JSplitPane split1;

    private String auditRecordLocation = "none";
    private AuditRecord auditRecord;
    private PersistedMvrManager mvrManager;
    private Boolean needsReading = true;

    private SortedManifest cardManifest;
    Map<String, StyleIF> poolMap = Collections.emptyMap();

    public CardTable(PreferencesExt prefs, TextHistoryPane infoTA, IndependentWindow infoWindow, float fontSize) {
        this.prefs = prefs;

        cardTable = new BeanTable<>(CardBean.class, (PreferencesExt) prefs.node("cardTable"), false,
                "CardManifest", "AuditableCard", null);
        cardTable.addListSelectionListener(e -> {
            CardBean cardBean = cardTable.getSelectedBean();
            if (cardBean != null) {
                setSelectedCard(cardBean);
            }
        });
        // poolTable.addPopupOption("Show Population", poolTable.makeShowAction(localInfo,
        //    bean -> ((PoolBean) bean).show()));

        setFontSize(fontSize);

        // layout of tables
        split1 = new JSplitPane(JSplitPane.VERTICAL_SPLIT, false, cardTable, localInfo);
        split1.setDividerLocation(prefs.getInt("splitPos1", 200));

        setLayout(new BorderLayout());
        add(split1, BorderLayout.CENTER);

        logger.debug("cardTable init");
    }

    public void setFontSize(float size) {
        cardTable.setFontSize(size);
        localInfo.setFontSize(size);
    }

    public boolean setAuditRecord(String auditRecordLocation) {
        logger.info("CardTable setAuditRecord " + auditRecordLocation);
        cardTable.setBeans(emptyList());

        this.auditRecordLocation = auditRecordLocation;
        AuditRecordIF auditRecord = AuditRecord.Companion.read(auditRecordLocation);
        if (auditRecord == null) {
            logger.info("CardTable failed on readFrom " + auditRecordLocation);
            return false;
        }
        if (auditRecord instanceof CompositeAuditRecord) return false;
        this.auditRecord = (AuditRecord) auditRecord;
        this.mvrManager = new PersistedMvrManager(this.auditRecord, false);

        needsReading = true;

        return true;
    }

    void setSelectedTab() {
        if (needsReading) {
            if (readCards()) needsReading = false;
        }
    }

    boolean readCards() {
        logger.info("start readCards for " + auditRecordLocation);

        Config config = auditRecord.getConfig();
        Integer cutoff = config.getRound().getSampling().getContestSampleCutoff();
        Integer ncardsToRead = (cutoff == null || cutoff < 11111) ? 11111 : cutoff;

        try {
            this.cardManifest = this.mvrManager.sortedManifest();

            List<StyleIF> styles = this.mvrManager.styles();
            if (styles != null) {
                Map<String, StyleIF> pools = new TreeMap<>(); // sorted
                for (StyleIF pool : styles) {
                    String cardStyle = "P" + pool.id();
                    pools.put(cardStyle, pool);
                }
                this.poolMap = pools;
            } else {
                List<CardPool> cardPools = this.mvrManager.pools();
                if (cardPools != null) {
                    Map<String, StyleIF> pools = new TreeMap<>(); // sorted
                    for (StyleIF pool : cardPools) {
                        String cardStyle = "P" + pool.id();
                        pools.put(cardStyle, pool);
                    }
                    this.poolMap = pools;
                }
            }

            List<CardBean> beanList = new ArrayList<>();
            int index = 1;
            try (var iter = cardManifest.getCards().iterator()) {
                while (iter.hasNext() && index < ncardsToRead) {
                    var card = iter.next();
                    beanList.add(new CardBean(card, index));
                    index++;
                }
            }
            cardTable.setBeans(beanList);
            logger.info("end readCards " + index + " cards from " + auditRecordLocation);

        } catch (Exception e) {
            e.printStackTrace();
            JOptionPane.showMessageDialog(null, e.getMessage());
            logger.error("setAuditRecord failed", e);
            return false;
        }

        return true;
    }

    StyleIF findPool(String cardStyle) {
        return poolMap.get(cardStyle);
    }

    void setSelectedCard(CardBean bean) {
        localInfo.setText(bean.show());
        localInfo.gotoTop();
    }

    public void saveState() {
        cardTable.saveState(false);

        prefs.putInt("splitPos1", split1.getDividerLocation());
    }

    //////////////////////////////////////////////////////////////////

    /* void showInfo(Formatter f) {
        if (this.auditRecord == null) { return; }
        if (this.auditConfig == null) { return; }
        if (this.auditConfig.getAuditType() != AuditType.ONEAUDIT) { return; }

        var cardPools = makeCardPoolsFromAuditRecord(auditRecord);
        String poolVotes = cardPools.showPoolVotes(4);
        f.format("%s", poolVotes);
    } */

    //////////////////////////////////////////////////////////////////

// data class AuditableCard (
//    val location: String, // info to find the card for a manual audit. Aka ballot identifier.
//    val index: Int,  // index into the original, canonical list of cards
//    val prn: Long,   // psuedo random number
//    val phantom: Boolean,
//    val possibleContests: IntArray, // remove
//
//    val votes: Map<Int, IntArray>?, // must have this and/or population
//    val poolId: Int?,
//    val cardStyle: String? = null, // remove
//    val population: StyleIF? = null, // not needed if hasStyle ?
//)

        // must be public
    public class CardBean {
        AuditableCardIF card;
        int index;

        public CardBean() {
        }

        CardBean(AuditableCardIF card, int index) {
            this.card = card;
            this.index = index;
        }

        public String getId() {
            return card.id();
        }
        public String getLocation() {
            return card.location();
        }

        public Integer getIndex() {
            return this.index;
        }

        public Integer getCardIndex() {
            return card.index();
        }

        public long getPrn() {
            return card.prn();
        }

        public Boolean getPhantom() {
            return card.phantom();
        }

        public String getContests() {
            int[] ids = card.possibleContests();
            StringBuilder sb = new StringBuilder();
            for (int id : ids) {
                sb.append("%d,".formatted(id));
            }
            return sb.toString();
        }

        public Integer getPoolId() {
            return card.poolId();
        }

        public String getCardStyle() {
            return card.styleName();
        }

        public String getPossibleContests() {
            var pop = card.style();
            int[] ids = pop.possibleContests();
            StringBuilder sb = new StringBuilder();
            for (int id : ids) {
                sb.append("%d,".formatted(id));
            }
            return sb.toString();
        }

        public String getVotes() {
            var votes = card.votes();
            if (votes == null) return "N/A";
            StringBuilder sb = new StringBuilder();
            SortedSet<Integer> ids = new TreeSet<>(votes.keySet());
            for (int contestId : ids) {
                sb.append("%d:".formatted(contestId));
                var cands = votes.get(contestId);
                if (cands.length == 0) sb.append(" ,");
                if (cands.length == 1) sb.append("%d, ".formatted(cands[0]));
                else if (cands.length > 1) {
                    sb.append("[");
                    for (int idx = 0; idx < cands.length; idx++) {
                        int cand = cands[idx];
                        if (idx > 0) sb.append(", ");
                        sb.append("%d".formatted(cand));
                    }
                    sb.append("],");
                }
            }
            return sb.toString();
        }

        public String show() {
            StringBuilder sb = new StringBuilder();
            sb.append(card.toString());
            sb.append("\n");

            var pop = card.style();
            sb.append(pop.toString());
            return sb.toString();
        }
    }
}

File: viewer/src/main/kotlin/org/cryptobiotic/rlauxe/viewer/CardTable.kt
/*
 * Copyright (c) 2026 John L. Caron
 * See LICENSE for license information.
 */

package org.cryptobiotic.rlauxe.viewer

import org.cryptobiotic.rlauxe.audit.AuditableCardIF
import org.cryptobiotic.rlauxe.audit.StyleIF
import org.cryptobiotic.rlauxe.persist.AuditRecord
import org.cryptobiotic.rlauxe.persist.AuditRecord.Companion.read
import org.cryptobiotic.rlauxe.persist.CompositeAuditRecord
import org.cryptobiotic.rlauxe.persist.SortedManifest
import org.cryptobiotic.rlauxe.workflow.PersistedMvrManager
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import ucar.ui.prefs.BeanTable
import ucar.ui.widget.IndependentWindow
import ucar.ui.widget.TextHistoryPane
import ucar.util.prefs.PreferencesExt
import java.awt.BorderLayout
import javax.swing.JOptionPane
import javax.swing.JPanel
import javax.swing.JSplitPane
import javax.swing.event.ListSelectionEvent
import javax.swing.event.ListSelectionListener

class CardTable(
    val prefs: PreferencesExt,
    val infoTA: TextHistoryPane,
    val infoWindow: IndependentWindow,
    fontSize: Float,
) : JPanel(), ViewerPanelIF {
    
    private val cardTable: BeanTable<CardBean>
    var localInfo: TextHistoryPane = TextHistoryPane()

    private val split1: JSplitPane

    private var auditRecordLocation: String? = "none"
    private var auditRecord: AuditRecord? = null
    private var mvrManager: PersistedMvrManager? = null
    private var needsReading = true

    private var cardManifest: SortedManifest? = null
    var poolMap: MutableMap<String, StyleIF> = mutableMapOf<String, StyleIF>()

    init {
        cardTable = BeanTable(
            CardBean::class.java, prefs.node("cardTable") as PreferencesExt, false,
            "CardManifest", "AuditableCard", null
        )
        cardTable.addListSelectionListener(ListSelectionListener { e: ListSelectionEvent? ->
            val cardBean = cardTable.getSelectedBean()
            if (cardBean != null) {
                setSelectedCard(cardBean)
            }
        })

        //cardTable.addPopupOption("Show Population", cardTable.makeShowAction(localInfo,
        //    bean -> ((cardTable) bean).show()));
        setFontSize(fontSize)

        // layout of tables
        split1 = JSplitPane(JSplitPane.VERTICAL_SPLIT, false, cardTable, localInfo)
        split1.setDividerLocation(prefs.getInt("splitPos1", 200))

        setLayout(BorderLayout())
        add(split1, BorderLayout.CENTER)

        logger.debug("cardTable init")
    }

    override fun setFontSize(size: Float) {
        cardTable.setFontSize(size)
        localInfo.setFontSize(size)
    }

    override fun setAuditRecord(auditRecordLocation: String): Boolean {
        logger.info("CardTable setAuditRecord " + auditRecordLocation)
        cardTable.setBeans(mutableListOf<CardBean?>())

        this.auditRecordLocation = auditRecordLocation
        val auditRecord = read(auditRecordLocation)
        if (auditRecord == null) {
            logger.info("CardTable failed on readFrom " + auditRecordLocation)
            return false
        }
        if (auditRecord is CompositeAuditRecord) return false
        this.auditRecord = auditRecord as AuditRecord
        this.mvrManager = PersistedMvrManager(this.auditRecord!!, false)

        needsReading = true

        return true
    }

    fun setSelectedTab() {
        if (needsReading) {
            if (readCards()) needsReading = false
        }
    }

    fun readCards(): Boolean {
        logger.info("start readCards for " + auditRecordLocation)

        val config = auditRecord!!.config
        val cutoff = config.round.sampling.contestSampleCutoff
        val ncardsToRead = if (cutoff == null || cutoff < 11111) 11111 else cutoff

        try {
            this.cardManifest = this.mvrManager!!.sortedManifest()

            val styles = this.mvrManager!!.styles()
            if (styles != null) {
                val pools = mutableMapOf<String, StyleIF>() // sorted
                for (pool in styles) {
                    val cardStyle = "P" + pool.id()
                    pools.put(cardStyle, pool)
                }
                this.poolMap = pools
            } else {
                val cardPools = this.mvrManager!!.pools()
                if (cardPools != null) {
                    val pools= mutableMapOf<String, StyleIF>() // sorted
                    for (pool in cardPools) {
                        val cardStyle = "P" + pool.id()
                        pools.put(cardStyle, pool)
                    }
                    this.poolMap = pools
                }
            }

            val beanList = mutableListOf<CardBean>()
            var index = 1
            cardManifest!!.cards.iterator().use { iter ->
                while (iter.hasNext() && index < ncardsToRead) {
                    val card = iter.next()
                    beanList.add(CardBean(card, index))
                    index++
                }
            }
            cardTable.setBeans(beanList)
            logger.info("end readCards " + index + " cards from " + auditRecordLocation)
        } catch (e: Exception) {
            e.printStackTrace()
            JOptionPane.showMessageDialog(null, e.message)
            logger.error("setAuditRecord failed", e)
            return false
        }

        return true
    }

    fun findPool(cardStyle: String?): StyleIF? {
        return poolMap.get(cardStyle)
    }

    fun setSelectedCard(bean: CardBean) {
        localInfo.setText(bean.show())
        localInfo.gotoTop()
    }

    override fun saveState() {
        cardTable.saveState(false)

        prefs.putInt("splitPos1", split1.getDividerLocation())
    }

    class CardBean(val card: AuditableCardIF, val sortedIndex: Int) {

        val id: String
            get() = card.id()

        val location: String
            get() = card.location()

        val manifestIndex: Int
            get() = card.index()

        val prn: Long
            get() = card.prn()

        val phantom: Boolean
            get() = card.phantom()

        val contests: String = card.possibleContests().contentToString()

        val poolId: Int?
            get() = card.poolId()

        val cardStyle: String
            get() = card.styleName()

        val possibleContests: String = card.style()?.possibleContests().contentToString() ?: ""

        val votes = buildString {
            if (card.votes() != null) {
                card.votes()!!.forEach { append("${it.key}:${it.value.contentToString()}, ") }
            }
        }

        fun show()= buildString {
            appendLine(card.toString())
            appendLine(card.style().toString())
        }

        companion object {
            @JvmStatic
            fun hiddenProperties() = "card";
        }

    }

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(CardPanel::class.java)
    }

}